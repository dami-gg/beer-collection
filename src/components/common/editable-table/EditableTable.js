// @flow
import React, { PureComponent } from "react";

import Row from "./Row";

import "./editable-table.scss";

class EditableTable extends PureComponent {
  props: {
    columns: Array<Object>,
    rows: Array<Object>,
    onEdit: Function,
    onCreate: Function,
    onDelete: Function
  };

  getHeader() {
    return this.props.columns.map((column: Object) => (
      <div className="editable-table__header__column" key={column.id}>
        {column.name}
      </div>
    ));
  }

  getRows() {
    return this.props.rows.map((row: Object, index: number) => (
      <Row
        row={row}
        type={index % 2 === 0 ? "light" : "dark"}
        key={row.id}
        columns={this.props.columns}
        onSave={this.props.onEdit}
        onDelete={this.props.onDelete}
      />
    ));
  }

  getExtraRow() {
    return <Row
        row={{}}
        type="extra"
        columns={this.props.columns}
        onSave={this.props.onCreate}
      />;
  }

  render() {
    return (
      <div className="editable-table">
        <div className="editable-table__header">
          {this.getHeader()}
        </div>
        {this.getExtraRow()}
        {this.getRows()}
      </div>
    );
  }
}

export default EditableTable;
