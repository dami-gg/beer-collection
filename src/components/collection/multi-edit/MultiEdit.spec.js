import React from "react";
import ConnectedMultiEdit, { MultiEdit } from "./MultiEdit";
import EditableTable from "../../common/editable-table/EditableTable";

import { shallow } from "enzyme";
import renderer from "react-test-renderer";
import configureStore from "redux-mock-store";

describe("MultiEdit", () => {
  let component,
    container,
    element,
    mock,
    storeConfig = configureStore();

  beforeEach(mocks);

  describe("dumb component", () => {
    it("should render without crashing", () => {
      renderDumbComponent();
    });

    it("should match the snapshot", () => {
      component = renderer.create(
        <MultiEdit collection={mock.state.collection} />
      );
      let json = component.toJSON();

      expect(json).toMatchSnapshot();
    });

    describe("render", () => {
      beforeEach(() => (component = renderDumbComponent()));

      it("should render an export button", () => {
        element = component.find(".multi-edit__export-button");

        expect(element.length).toEqual(1);
      });

      it("should render a EditableTable", () => {
        element = component.find(EditableTable);

        expect(element.length).toEqual(1);
      });
    });
  });

  describe("smart component", () => {
    it("should render without crashing", () => {
      renderSmartComponent();
    });

    describe("connection to Redux store", () => {
      beforeEach(() => {
        container = renderSmartComponent();
      });

      it("should match props with state", () => {
        expect(container.prop("collection")).toEqual(mock.state.collection);
      });
    });
  });

  function renderDumbComponent() {
    return shallow(<MultiEdit />);
  }

  function renderSmartComponent(initialState = mock.state) {
    let store = storeConfig(initialState);

    return shallow(<ConnectedMultiEdit store={store} />);
  }

  function mocks() {
    mock = {
      state: {
        collection: []
      }
    };
  }
});
