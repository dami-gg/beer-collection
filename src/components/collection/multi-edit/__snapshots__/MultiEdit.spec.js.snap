// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiEdit dumb component should match the snapshot 1`] = `
<div
  className="multi-edit"
>
  <button
    className="button button--blue multi-edit__export-button"
    onClick={[Function]}
    type="button"
  >
    Export as CSV file
  </button>
  <div
    className="editable-table"
  >
    <div
      className="editable-table__header"
    >
      <div
        className="editable-table__header__column "
      >
        Name
      </div>
      <div
        className="editable-table__header__column "
      >
        Type
      </div>
      <div
        className="editable-table__header__column "
      >
        Origin
      </div>
      <div
        className="editable-table__header__column "
      >
        Rating
      </div>
      <div
        className="editable-table__header__column mobile-hidden"
      >
        Image
      </div>
    </div>
    <div
      className="row row--edit"
    >
      <div
        className="cell "
      >
        <input
          className="cell__input cell__input--text"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
      <div
        className="cell "
      >
        <input
          className="cell__input cell__input--text"
          onChange={[Function]}
          placeholder="Type"
          type="text"
          value=""
        />
      </div>
      <div
        className="cell "
      >
        <input
          className="cell__input cell__input--text"
          onChange={[Function]}
          placeholder="Origin"
          type="text"
          value=""
        />
      </div>
      <div
        className="cell "
      >
        <input
          className="cell__input cell__input--number"
          onChange={[Function]}
          placeholder="Rating"
          type="text"
          value=""
        />
      </div>
      <div
        className="cell mobile-hidden"
      >
        <div
          className="cell__image-upload"
        >
          <div
            className="image-uploader"
          >
            <div
              className="image-uploader__preview 
                      image-uploader__preview--empty
                      image-uploader__preview--hidden"
            />
            <div
              className="image-uploader__uploader"
            >
              <button
                className="button button--blue image-uploader__button"
                onClick={[Function]}
                type="button"
              >
                <label
                  className="image-uploader__button__label"
                  htmlFor="image-uploader__input"
                >
                  Select image
                </label>
              </button>
              <input
                className="image-uploader__input"
                id="image-uploader__input"
                onChange={[Function]}
                type="file"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        className="buttons"
      >
        <button
          className="button button--green "
          onClick={[Function]}
          type="button"
        >
          Save
        </button>
        <button
          className="button button--grey "
          onClick={[Function]}
          type="button"
        >
          Cancel
        </button>
      </div>
    </div>
  </div>
</div>
`;
